### PR TITLE
Add Tuya button support

### PR DESCRIPTION
**Pull request by dschall**
This change adds support for various Tuya / Zemismart battery-operated switches.
These devices send custom Zigbee commands (ID: 253) and expose duplicate PowerConfiguration endpoints.

### CODE REVIEW

File: org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+/**
+ * Generic converter for Tuya buttons (e.g., Zemismart).
+ * <p>
+ * This converter needs to be configured with the ZigBee commands that are triggered by the button presses. This is done
+ * by channel properties that specify the endpoint, the cluster, the command ID, and (optionally) a command parameter.
+ * <p>
+ * As the configuration is done via channel properties, this converter is usable via static thing types only.
+ *
+ * @author Henning Sudbrock - initial contribution
+ * @author Thomas Weißschuh - support for attribute-based buttons
```
Comment (dschall) [inline org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java:50]: Update authors

Comment (dschall) [reply org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java:50]: .. and comment

File: org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
```
Comment (dschall) [inline org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java:2]: Update year

File: org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/command/TuyaButtonPressCommand.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
```
Comment (dschall) [inline org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/command/TuyaButtonPressCommand.java:2]: Update year

File: org.openhab.binding.zigbee/src/main/resources/OH-INF/thing/tuya/ts0044.xml [hunk 4]
```diff
@@ -0,0 +1,76 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="zigbee"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="tuya_ts0044" listed="false">
```
Comment (dschall) [inline org.openhab.binding.zigbee/src/main/resources/OH-INF/thing/tuya/ts0044.xml:7]: I think the button mapping for the 4-button switch is incorrect, Thomas (@oggerschummer) can provide the correct mapping.

File: org.openhab.binding.zigbee/src/main/resources/discovery.txt [hunk 5]
```diff
@@ -11,3 +11,7 @@ xiaomi_lumisensor-switchaq2,modelId=lumi.sensor_switch.aq2
 xiaomi_lumisensorwaterleak,modelId=lumi.sensor_wleak.aq1
 innr-rc-110,vendor=innr,modelId=RC 110
 osram-switch-4x-eu,vendor=OSRAM,modelId=Switch 4x EU-LIGHTIFY
+tuya_ts0041,vendor=_TZ3000_tk3s5tyg,modelId=TS0041
```
Comment (dschall) [inline org.openhab.binding.zigbee/src/main/resources/discovery.txt:14]: We could remove the vendor property or change this to a regex match, e.g. _TZ3000_\w{8} .

**Comment by dschall** [comment]
todo: add tests

**Comment by cdjackson** [comment]
Thanks @dschall - at the very top level, I think the approach looks fine - thanks for looking at this. I've only quickly scanned the code to check a few things I had in mind that could be problematic and these look ok, so I will take a better look in a few days.

File: org.openhab.binding.zigbee/src/main/resources/OH-INF/thing/tuya/ts0044.xml (hunk 4 above)
Comment (Oggerschummer) [reply org.openhab.binding.zigbee/src/main/resources/OH-INF/thing/tuya/ts0044.xml:7]: Just checked with my second switch. The ordering you have is correct. The first switch shipped with an error in the manual where the buttons where numbered differently.
No change required.

**Comment by dschall** [comment]
> Thanks @dschall - at the very top level, I think the approach looks fine - thanks for looking at this. I've only quickly scanned the code to check a few things I had in mind that could be problematic and these look ok, so I will take a better look in a few days.

Hey @cdjackson, I've added tests and fixed a bug. I think now the PR is ready to merge.
I would appreciate comments.

**Comment by dschall** [comment]
@cdjackson Hey Chris, can we merge this?

**Comment by cdjackson** [comment]
Sorry - I have limited time right now and probably for about the next 2 weeks. Due to travel restrictions we have to move house in very short time.

**Comment by dschall** [comment]
@cdjackson, gentle ping to please review this PR :)

**Comment by suppressio** [comment]
Hi, what's the state of this project?

**Comment by suppressio** [comment]
@dschall Hi... I'm interested too... another gentle ping to review this PR ...

**Comment by cdjackson** [comment]
Sorry - I'm in the process of moving and have limited time. I will try and look at this in the next few weeks as things hopefully calm down. I will likely refactor this into a separate bundle.

**Comment by dschall** [comment]
@cdjackson, gentle ping.
Are you going to rewrite this into a separate bundle, or should I resolve the conflicts and complete this PR (assuming you'd sign off on it).

**Comment by gauthier12** [comment]
Hi. I'm interested too. Maybe it is better to resolve the conflicts and merge instead of losing the development.
It is possible to refactor into a separate bundle later

**Comment by deeja** [comment]
@cdjackson  - would you have any time to look at this again? I just got one of these switches and pretty excited to get it running! Will shout you a beer :D

**Comment by cdjackson** [comment]
Sorry for the delay with this - I will try and take a look at this over the next few days.

File: org.openhab.binding.zigbee/src/main/resources/discovery.txt [hunk 6]
```diff
@@ -15,4 +15,7 @@ xiaomi_lumisensormagnet,modelId=lumi.sensor_magnet
 xiaomi_lumiremotemini,modelId=lumi.remote.b1acn01
 innr-rc-110,vendor=innr,modelId=RC 110
 osram-switch-4x-eu,vendor=OSRAM,modelId=Switch 4x EU-LIGHTIFY
-legrand_dimmer_without_neutral,vendor=\u001f Legrand,modelId=\u001f Dimmer switch w/o neutral
```
Comment (dschall) [inline org.openhab.binding.zigbee/src/main/resources/discovery.txt:18]: oh dang it, I think I accidentally removed that line :(

File: org.openhab.binding.zigbee/src/main/resources/discovery.txt (hunk 5 above)
Comment (dschall) [reply org.openhab.binding.zigbee/src/main/resources/discovery.txt]: removed the vendor instead. modelId should be sufficient.

File: org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        {
+            logger.debug("{} ignoring duplicate command {}", endpoint.getIeeeAddress(), thisTxId);
+        }
+        else if (command instanceof TuyaButtonPressCommand) {
+            TuyaButtonPressCommand tuyaButtonPressCommand = (TuyaButtonPressCommand) command;
+           thing.triggerChannel(channel.getUID(), getEventType(tuyaButtonPressCommand.getPressType()));
+           clientCluster.sendDefaultResponse(command, ZclStatus.SUCCESS);
+        }
+        else {
+           logger.warn("{} received unknown command {}", endpoint.getIeeeAddress(), command);
```
Comment (cdjackson) [inline org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java:189]: Being petty, but please addd a `:` after the `{}` for the address - it's good to keep things standard as sometimes processing of logs looks for certain features.

File: org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    @Override
+    public Channel getChannel(ThingUID thingUID, ZigBeeEndpoint endpoint) {
+        // This converter is used only for channels specified in static thing types, and cannot be used to construct
+        // channels based on an endpoint alone.
+        return null;
+    }
+
+    @Override
+    public boolean commandReceived(ZclCommand command) {
+        logger.debug("{} received command {}", endpoint.getIeeeAddress(), command);
```
Comment (cdjackson) [inline org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java:177]: Being petty, but please addd a `:` after the `{}` for the address - it's good to keep things standard as sometimes processing of logs looks for certain features.

File: org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+        switch(pressType)
+        {
+            case 0:
+                return CommonTriggerEvents.SHORT_PRESSED;
+            case 1:
+                return CommonTriggerEvents.DOUBLE_PRESSED;
+            case 2:
+                return CommonTriggerEvents.LONG_PRESSED;
+            default:
+                logger.warn("{} received unknown pressType {}", endpoint.getIeeeAddress(), pressType);
```
Comment (cdjackson) [inline org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java:212]: Being petty, but please addd a `:` after the `{}` for the address - it's good to keep things standard as sometimes processing of logs looks for certain features.

**Comment by cdjackson** [comment]
@dschall firstly, sorry for the slight pettiness on the logging format but I like to try and keep things standard so any searches, or log viewer type processing has a consistent format to look for...

Otherwise I'm happy enough with this - it all looks good so if you can make those couple of small changes I'll merge.

File: org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java (hunk 8 above)
Comment (dschall) [reply org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java]: done, also, the others. I'll resolve without adding a separate comment to each.

Comment (cdjackson) [reply org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java]: Thanks.

**Pull request merged**